#15979: Switch to google benchmark for pgm dispatch tests

jbaumanTT opened the pull request:
### Ticket
#15979

### Problem description
The current output format works well with a smaller number of tests, but with a large number of tests it's hard to connect the resulting number to the test that created it. This particular hinders storing test results in a database and comparing them across runs.


### What's changed
Switch to using the google benchmark framework, which outputs a JSON file containing the results of all tests. By default, running the binary runs all the benchmarks from sweep_pgm_dispatch.sh. The set of tests to be run can be filtered using the `--benchmark_filter=<regex>` command-line argument.

One-off test cases can be run by passing `--custom` and the command line arguments as before.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

jbaumanTT committed:
#16547: Switch to google benchmark for pgm dispatch tests

The current output format works well with a smaller number of tests, but with a
large number of tests it's hard to connect the resulting number to the test
that created it. This particular hinders storing test results in a database and
comparing them across runs.

Switch to using the google benchmark framework, which outputs a JSON file
containing the results of all tests. By default, running the binary runs all
the benchmarks from sweep_pgm_dispatch.sh. The set of tests to be run can be
filtered using the `--benchmark_filter=<regex>` command-line argument.

One-off test cases can be run by passing `--custom` and the command line arguments as before.

pgkeller reviewed (commented):
I use filt_pgm_dispatch for the bw_and_latency test as well, so don't delete that.
I still want to dump these results to a spreadsheet, what does that flow look like?
Otherwise looks good

jbaumanTT commented:
> I use filt_pgm_dispatch for the bw_and_latency test as well, so don't delete that. I still want to dump these results to a spreadsheet, what does that flow look like? Otherwise looks good

Ok, added back filt_pgm_dispatch.pl. I've added a new json_to_csv.py that can be used to dump to a CSV (one line per test). The workflow is to run
```
build/test/tt_metal/perf_microbenchmark/dispatch/test_pgm_dispatch --benchmark_out_format=json --benchmark_out=bench.json
tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/json_to_csv.py bench.json > bench.csv
```
google benchmark framework also has a native way to output CSVs, but it's deprecated and requires some extra work.

This patch also uploads the json file from the CI bots, so it's pretty easy to download that and convert it.

tt-rkim commented on .github/workflows/fast-dispatch-frequent-tests-impl.yaml:
```diff
@@ -57,3 +57,9 @@ jobs:
+        with:
+          path: |
+            bench.json
+          prefix: "bench_json_"
```
A little generic. `pgm_benchmarks_json_` ?

jbaumanTT replied:
Done.

tt-rkim commented on tests/tt_metal/tt_metal/perf_microbenchmark/CMakeLists.txt:
```diff
@@ -75,4 +75,6 @@ foreach(TEST_SRC ${PERF_MICROBENCH_TESTS_SRCS})
     list(APPEND PERF_MICROBENCH_TEST_TARGETS ${TEST_TARGET})
 endforeach()
 
+target_link_libraries(test_pgm_dispatch PUBLIC benchmark::benchmark)
```
```suggestion
target_link_libraries(test_pgm_dispatch PRIVATE benchmark::benchmark)
```

jbaumanTT replied:
Done.

tt-rkim commented on tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/compare_pgm_dispatch_perf_ci.py:
```diff
@@ -0,0 +1,70 @@
+import sys
+
+os.chdir(os.getenv("TT_METAL_HOME"))
+golden = json.load(open("tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.json", "r"))
```
Hmm... could we separate the call of the test and the post-processing? So you don't have to worry about this `os.system` business? This file could just take in the `bench.json` output

tt-rkim replied:
Then you don't have to worry about handling errors in here etc

tt-rkim replied:
You could make this a separate step in the workflow YAML file

jbaumanTT replied:
Ok, moved this out and into the workflow yaml file. I kept both commands as part of the same step, since it didn't make sense to me to invoke docker separately for each.

tt-rkim commented on .github/workflows/fast-dispatch-frequent-tests-impl.yaml:
```diff
@@ -16,15 +16,13 @@ jobs:
-              timeout: 10
-            },
-          ]
+          - name: "WH N300 pgm dispatch nightly"
```
Haha you like native YAML, eh?

jbaumanTT replied:
It's so much prettier...

tt-rkim commented on tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/compare_pgm_dispatch_perf_ci.py:
```diff
@@ -0,0 +1,75 @@
+golden = json.load(
+    open(
+        os.path.join(
+            os.getenv("TT_METAL_HOME"), "tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.json"
```
Just a small nit - for future proofing, we'd like to reduce usage of `TT_METAL_HOME`

In Python, you can instead resolve the full path of `tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/compare_pgm_dispatch_perf_ci.py` using `__file__` and use that to find out where the JSON golden is.

jbaumanTT replied:
Ok, done.

tt-rkim commented on tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/json_to_csv.py:
Hmm is this used anywhere now?

jbaumanTT replied:
We just want to use it locally to help copying the benchmark results into spreadsheets.

tt-rkim commented:
@TT-billteng LOOK SEE LOOK SEE
GOOGLE BENCHMARK!!!
